mantidproject/mantid
# Update Clang to version 18 on macOS

The version of `clang` that `conda-forge` uses is now 18, see [here](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/e6c41e91ba8a8f4f2edb087f5e8be6428c6c3374/recipe/conda_build_config.yaml#L30).

`llvm-openmp` was also updated to version 18 for consistency.

Fixes #38335. 

As well as matching the version of `clang` on `conda-forge`, `libc++` no longer supports `clang` version 16, see [here](https://libcxx.llvm.org/).

To fix a problem casting an `IValidator` to an `ArrayBoundedValidator<double>` we had to disable an optimisation introduced in `clang` 17 with the `-fno-assume-unique-vtables` flag. See the release notes [here](https://releases.llvm.org/17.0.1/tools/clang/docs/ReleaseNotes.html#what-s-new-in-clang-release).

We also disabled compiler extensions to help maintain code compatibility between different compilers.




### To test:

- Remove `build/CMakeCache.txt`
- Create a new conda environment
- Install a local build of the `mantid-developer` metapackage from this branch
- Hit compile and go for a cup of tea

Double check that the CI used the correct `clang` version.

For the macOS system tests:

https://builds.mantidproject.org/job/build_packages_from_branch/948/







---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
